internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -395,70 +396,101 @@ public static ControlMessage generateHeartbeatMessage(CheckSumType checkSumType)
    */
   @Override
   public void close(boolean gracefulClose) {
+    close(gracefulClose, true);
+  }
+
+  public void close(boolean gracefulClose, boolean retryOnGracefulCloseFailure) {
     synchronized (closeLock) {
       if (isClosed) {
         return;
       }
-      long startTime = System.currentTimeMillis();
-      logger.info("Closing VeniceWriter for topic: {}", topicName);
-      try {
-        // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
-        // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
-        if (gracefulClose) {
-          endAllSegments(true);
-        }
-        // DO NOT call the {@link #PubSubProducerAdapter.close(int) version from here.}
-        // For non-shared producer mode gracefulClose will flush the producer
 
-        producerAdapter.close(topicName, closeTimeOutInMs, gracefulClose);
-        OPEN_VENICE_WRITER_COUNT.decrementAndGet();
-      } catch (Exception e) {
-        logger.warn("Swallowed an exception while trying to close the VeniceWriter for topic: {}", topicName, e);
-        VENICE_WRITER_CLOSE_FAILED_COUNT.incrementAndGet();
+      logger.info("Closing VeniceWriter for topic: {}, gracefulness: {}", topicName, gracefulClose);
+      try (Timer ignore = Timer.run(
+          elapsedTimeInMs -> logger.info("Closed VeniceWriter for topic: {} in {} ms", topicName, elapsedTimeInMs))) {
+        try {
+          // If {@link #broadcastEndOfPush(Map)} was already called, the {@link #endAllSegments(boolean)}
+          // will not do anything (it's idempotent). Segments should not be ended if there are still data missing.
+          if (gracefulClose) {
+            endAllSegments(true);
```

Comment:
https://github.com/linkedin/venice/blob/main/internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java#L1693
Sending control messages to a non-existing topic can stuck forever since this function is calling `get()` without timeout.

If we don't address this logic, the fallback can only execute in the application logic to have a timer and ungraceful close the writer.

Reply:
Another strategy to this problem:
can we always close the VeniceWriter ungracefully?

Reply:
There is one more way to think about to address this stuck consumer issue:
1. Check the thread trace of the stuck consumer to see whether it is blocked by Producer#send.
2. If yes, check the producing topic exist or not.
3. If not, kill the corresponding VeniceWriter.

This will be a little hacky, but it would catch all the cases of stuck producers.